Deleted file, fix trigger and added test

malgorzataKraw opened the pull request:
1. Deleted not needed files with old triggers and test class
2. Fix triggers and added class handler, added factory class for test for telemetries
3. Added test for trigger for CarPosting ,added factory class for CarPosting and CarsDealership

Commit:
Remove not needed files

Commit:
Fix trigger with addded class handler, added factory class for tests for async apex class

Commit:
Added test for trigger and added class factory

kingabadura commented on force-app/main/default/classes/BatchTelemetriesDeletionTest.cls:
```diff
@@ -0,0 +1,61 @@
+@isTest
```
Każda klasa i metoda powinna mieć javadoki, w tym klasy testowe i triggery. (Nie będę dopisywać już tego dalej, ale dotyczy wszystkich plików w MR)

kingabadura commented on force-app/main/default/classes/BatchTelemetriesDeletionTest.cls:
```diff
@@ -0,0 +1,61 @@
+@isTest
+public class BatchTelemetriesDeletionTest {
+    @isTest static void testBatchWhenInsertOldTelemetries () {
+        
```
Usunąć entery

kingabadura commented on force-app/main/default/classes/BatchTelemetriesDeletionTest.cls:
```diff
@@ -0,0 +1,61 @@
+@isTest
+public class BatchTelemetriesDeletionTest {
+    @isTest static void testBatchWhenInsertOldTelemetries () {
```
raczej annotacje dajemy nad metodę (konwencja)

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+public class CarHandler {
+    
+    
+    // this method have to be changed
```
Usunąć zbędne komentarze

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+        }
+        
+        
+        
```
Baaaardzo dużo enterów !!! :D Proszę ogranicz entery tylko do miejsc gdzie są **niezbędne** do zachowania czytelności. Metody powinny być krótkie i zwięzłe, bez zbędnych spacji i enterów, dobrze sformatowane.

kingabadura commented on force-app/main/default/classes/BatchTelemetriesDeletionTest.cls:
```diff
@@ -0,0 +1,61 @@
+        Datetime dayBeforeYesterday = Datetime.now().addDays(-2);
+        List<Telemetry__c> lst_telemetries = TestTelemetriesFactory.createTelemetryInsert(100,dayBeforeYesterday);
+        
+        System.debug('List with telemetries, size before deleted old records: '+lst_telemetries.size());
```
Unikamy zbędnych debugów.

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+    
+    
+    // this method have to be changed
+    public static void carDeletion(List<Car__c> carList1){
```
czemu carList**1**? czemu nie carList :D? albo jeszcze lepiej carsToDelete albo coś w tym rodzaju?

kingabadura replied:
W metodach publicznych sprawdzamy parametry wejściowe

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+            insert newReview;
+            
+        }
+        
```
W momencie usuwania pojazdu powinny być usuwane równiez dane telemetryczne.

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+        List<ReviewCarDeletion__c> newReview = new List<ReviewCarDeletion__c>();
+        
+        for (Car__c c :carList1){
+            ReviewCarDeletion__c rcd = new ReviewCarDeletion__c(Name='Raport '+c.Name);
```
Nie hardkodujemy stringów, wszystkie teksty wyciągamy do stałych.

kingabadura replied:
Nie stosujemy skrótowych nazw dla zmiennych, zgodnie z Good Practisami. Nie piszemy acc.. tylko accountToUpdate, nie piszemy c tylko case, nie piszemy avg tylko avarage itp

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+        
+        
+        if(newReview.size()>0){
+            
```
Entery

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+            
+            rcd.ReportDeletion__c = 'Deletion car with name '+c.Name +' with brand '+c.Brand__c;
+            newReview.add(rcd);
+            
```
Entery

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+        
+        List<ReviewCarDeletion__c> newReview = new List<ReviewCarDeletion__c>();
+        
+        for (Car__c c :carList1){
```
Nazwa zmiennej, nie stosujemy skrótów!

kingabadura commented on force-app/main/default/classes/CarHandler.cls:
```diff
@@ -0,0 +1,32 @@
+    // this method have to be changed
+    public static void carDeletion(List<Car__c> carList1){
+        
+        List<ReviewCarDeletion__c> newReview = new List<ReviewCarDeletion__c>();
```
Nazwa zmiennej.. Listy dobrze żeby miały w nazwie List, np reviewsList, accountsList itp. Mapy dobrze żeby miały w nazwie na końcu Map, sety Set itp itd

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+public class CarPostingHandler {
+    
+    public static void addCarPostingPhysical(List<Car_Posting__c> newCarPost1){
```
Nazwa zmiennej List<Car_Posting__c> newCarPost1

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+                
+                Id idCar = c.Car__c;
+                
+                List<Car_Posting__c> carList = [Select id, Car__c, CarsDealership__c From Car_Posting__c where Type__c='Physical' and Car__c =: idCar];
```
**SELECT w pętli !!! Absolutnie niedozwolone jest robić selectów i DML w pętlach**

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+                if(carList.size()>0){
+                    
+                    Id carsDeal = carList.get(0).CarsDealership__c;
+                    CarsDealership__c carD = [select id, name from CarsDealership__c where id =: carsDeal Limit 1];
```
Kolejny select w pętli

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+                    
+                    Id carsDeal = carList.get(0).CarsDealership__c;
+                    CarsDealership__c carD = [select id, name from CarsDealership__c where id =: carsDeal Limit 1];
+                    c.AddError('Only one physical offer could exist! This car is already in CarDealership : '+carD.Name);
```
Hardcode

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+        
+        Id idCarDealer = c.CarsDealership__c;
+        
+        CarsDealership__c cd = [SELECT Id,Car_Quantity__c FROM CarsDealership__c WHERE 
```
Selecty selecty.. w pętli.

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+    }
+    
+    
+    public static void addCarPostingLimit(List<Car_Posting__c> newCarPost2){
```
nazwa zmiennej..  nazwy zmiennych w metodach **nie** muszą być unikatowe w skali klasy, przez co rozumiem że spokojnie parametr w każdej metodzie w ramach jednej klasy **może nazywać się tak samo**! Nie musisz dodawać 1,2,3..

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+        for (Car_Posting__c c : newCarPost1){
+            
+        
+            if(c.Type__c == 'Physical'){
```
Hardcode

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+        
+        CarsDealership__c cd = [SELECT Id,Car_Quantity__c FROM CarsDealership__c WHERE 
+                                                id =: idCarDealer LIMIT 1];
+        Decimal carQuant = cd.Car_Quantity__c;
```
Przed użyciem zmiennej w zapytaniu trzeba sprawdzić czy ona ma wartość

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+        
+       
+        if ((carPostListExist.size()+1) > carQuant){
+            c.AddError('Available only '+carQuant+' car posting!');
```
Hardcode.. nie będę dopisywać już w kolejnych, ale prośba o zamianę wszystkich stringów na zmienne/stałe lub labelki

kingabadura commented on force-app/main/default/triggers/CarPostingTrigger.trigger:
```diff
@@ -0,0 +1,26 @@
+        if(Trigger.isInsert){
+            
+            CarPostingHandler.addCarPostingPhysical(Trigger.New);
+            
```
Entery

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+    
+    public static void sendEmailSalesDir(List<Car_Posting__c> newCarPost3){
+        
+        UserRole ur = [SELECT Id FROM UserRole WHERE Name='Director, Direct Sales' LIMIT 1];
```
Nazwy zmiennych

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+        String subject = 'Information about new car postings';
+        String body = '';
+        
+        System.debug('from method sendEmailSalesDir');
```
Debug.. jeśli celowo chcemy żeby w metodzie był debug, to piszemy je zawsze stosując Logging.Level.

kingabadura commented on force-app/main/default/classes/CarPostingHandler.cls:
```diff
@@ -0,0 +1,79 @@
+            body = body+'New carPosting with name: '+c.Name + ' was added. With type car posting '+c.Type__c+' . ';
+        }
+        System.debug('Body email: '+body);
+        System.debug('Email sent to: '+ emailDir + ' with subject: '+subject);
```
Usunąć zbędne debugi. Debugi mają sens :
- przy wejściu do metody
- jeśli kończymy wykonywanie metody bo parametry są błedne lub z jakiegoś innego powodu
- przy operacjach DML
- przy errorach w catchu

I zawsze powinny mieć ustawiony logging level

kingabadura commented on force-app/main/default/triggers/CarsDealershipTrOld.trigger:
```diff
@@ -33,9 +33,8 @@ trigger CarToWarehouse1 on CarsDealership__c (after update) {
         List<Car_Posting__c> carPostInWarehouse = new List <Car_Posting__c>();
         
-        // Relation: car dealership <- car posting -> car
+        // car dealership <- car posting -> car
```
Usunąć komentarze

kingabadura commented on force-app/main/default/classes/CarPostingTriggerTest.cls:
```diff
@@ -0,0 +1,278 @@
+        
+        insert acct;
+
+        // Previously I try use TestCarPostingFactory and CarsDealershipFactory
```
Komentarze do usunięcia

Commit:
Provided changes per comments:

- deleted enters, comments, debug
- rename methods, fiels and object
- changes methods in handler without using select in loops
- added class to create renovation Case (task 5)

Commit:
Changed  tests, added CarFactory and AccountFactory class, changed trigger for car posting, added telemetry deletion with deleted car

kingabadura commented on force-app/main/default/classes/AccountFactory.cls:
```diff
@@ -0,0 +1,35 @@
+/**
+* @author name <malgorzata.krawczyk@accenture.com>
```
`@author name <malgorzata.krawczyk@accenture.com>`
to name to celowe :D?

kingabadura commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,43 @@
+     * whether there are free renovation teams
+     **/
+    public void execute (QueueableContext context){
+        CarsDealership__c carsDealership = [Select Id, Name From CarsDealership__c Where Id =:carDealershipId LIMIT 1];
```
Selecty zgodnie z good practice zawsze robimy do listy :) Nawet jeśli oczekujemy jednego elementu. Jest tak dlatego, że jeśli ten select nic nie zwróci poleci nam exception, selectując do listy dostaniemy po prostu pustą listę.

kingabadura commented on force-app/main/default/classes/CarsDealershipHandler.cls:
```diff
@@ -0,0 +1,48 @@
+                for (CarsDealership__c carsDealershipNew : carsDealershipNewList){
+                    if(carsDealershipNew.Dealership_Status__c == carDealershipStatusClose && carsDealershipNew.id == carsDealershipOld.id ){
+                        AsyncCaseRenovation asyncCaseRenovation = new AsyncCaseRenovation(carsDealershipNew.id);
+                        Id jobID = System.enqueueJob(asyncCaseRenovation);   
```
W tym fragmencie jest conajmniej kilka problemów:
1) Po pierwsze.. zawsze zwracaj uwagę na limity.. Możemy wywołać maks 50 queueable jobów.. jeśli wpadnie nam tu więcej niż 50 rekordów to niestety ta logika zawiedzie.
`for (CarsDealership__c carsDealershipNew : carsDealershipNewList)`
2) For w forze.. 
```
for (CarsDealership__c carsDealershipOld : carsDealershipOldList){
            if(carsDealershipOld.Dealership_Status__c == carDealershipStatusOpen){

                for (CarsDealership__c carsDealershipNew : carsDealershipNewList){
```
Bardzo, oj bardzo niewydajne rozwiązanie.. złożoność obliczeniowa takiego rozwiązania jest ogromna! A wystarczyłoby użyć Map które mamy out of the box w triggerze, w szczególności Trigger.oldMap i po ID wyciągnąć "starą" wersję obiektu, zamiast forem przeszukiwać obie mapy ;)... 
3) To już pierdółka, ale po co ci zmienna jobID?
`Id jobID = System.enqueueJob(asyncCaseRenovation);`

kingabadura commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,43 @@
+     **/
+    public void execute (QueueableContext context){
+        CarsDealership__c carsDealership = [Select Id, Name From CarsDealership__c Where Id =:carDealershipId LIMIT 1];
+        String subjectRenovation = 'Renovation Case';
```
Jeśli chodzi o te Stringi to lepsza by była klasa do constantów, coby w razie zmiany nazwy statusu, subjectu itp. nie musieć szukać po całym projekcie zmiennych z nazwą subjectu tylko zmienić w jednym miejscu ;) Tyczy się wszystkich stringów. Ewentualnie nada się też label lub chociaż zmienna final w klasie, nie w metodzie..

kingabadura commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,43 @@
+        String subjectRenovation = 'Renovation Case';
+        String subjectCase = subjectRenovation + carsDealership.Name;
+
+        Case newCase = new Case( Subject = subjectCase, Status='New', Origin= 'Phone',CarsDealershipRel__c = carDealershipId);
```
Hardcody - status, origin itp

kingabadura commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,43 @@
+        if(response.getStatusCode() == 201) {
+            newCase.Status = 'Closed';
+        } else {
+            newCase.Status = 'Failed';
```
Hardcody, Closed & Failed

kingabadura commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,43 @@
+     * @author name <malgorzata.krawczyk@accenture.com>
+     * @date 26/08/2021
+     * @description The method is preparing a new renovation case and sending an inquiry to the external renovation system to check
+     * whether there are free renovation teams
```
Idealnie jakby w tych javadokach znalazły się jeszcze parametry @param i typ zwracany @return

kingabadura commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,43 @@
+/**
+* @author name <malgorzata.krawczyk@accenture.com>
```
Brak klasy testowej

kingabadura commented on force-app/main/default/classes/BatchTelemetriesDeletionTest.cls:
```diff
@@ -0,0 +1,65 @@
+* @description This class has test methods for the batch class
+**/
+@isTest
+public class BatchTelemetriesDeletionTest {
```
`BatchTelemetriesDeletionTest `
Raczej konwencja jest taka, że nazwa klasy testowej jest taka sama jak testowanej +Test na końcu..
W tym wypadku szukam po kodzie gdzie jest klasa  BatchTelemetriesDeletion i nie widzę jej, ciężko dojść która klasa testowa co testuje "na pierwszy rzut oka".. również pipeline na projektach nie ogarnia jeśli nazwy są inne. Prośba o utrzymanie konwencji.

kingabadura commented on force-app/main/default/classes/OldTelemetriesDeletion.cls:
```diff
@@ -0,0 +1,36 @@
+* @date 26/08/2021
+* @description This class implements Batchable and applies to data that was created the day before yesterday
+**/
+public class OldTelemetriesDeletion implements Database.Batchable<sObject>, Database.Stateful{
```
Fajnie jakby klasa która jest Batchem miała Batch na końcu, Schedulablem - Schedule/Scheduler, Queueable - Queueable itp. Poprawia to czytelność i od razu wiadomo co dana klasa robi.

kingabadura commented on force-app/main/default/classes/AccountFactory.cls:
```diff
@@ -0,0 +1,35 @@
+     * @description This method creates one account record
+     **/
+    public static Account getSimpleAccount(){
+        String accountName = 'Test Account';
```
Domyślam się, że to miało na celu uniknięcie hardkokodwania, ale akurat w tym przypadku to nie jest konieczne :D 
` String accountName = 'Test Account';`

Commit:
Changed javadoc and added class with constants

Commit:
Changed method with close car dealership

mtargiel commented on force-app/main/default/classes/AccountFactory.cls:
```diff
@@ -0,0 +1,38 @@
+        List<Account> accountList = new List<Account>();
+
+        for (Integer i = 0; i < numberAccounts; i++) {
+            Account account = new Account (Name = Constants.ACCOUNT_TEST_NAME + i); 
```
A może użyj getSimpleAccount?

mtargiel commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,47 @@
+     * @Task: Task 5
+     **/
+    public void execute (QueueableContext context){
+        List<CarsDealership__c> carsDealershipList = new List<CarsDealership__c>([Select Id, Name From CarsDealership__c Where Id =:carDealershipId]);
```
Co jeżeli nie znajdzie nic?

mtargiel commented on force-app/main/default/classes/AsyncCaseRenovation.cls:
```diff
@@ -0,0 +1,47 @@
+
+        Http server = new Http();
+        HttpRequest request = new HttpRequest();
+        request.setEndpoint('https://jsonplaceholder.typicode.com/posts');
```
endpointa nie hardkodujemy, użyj named credential

Commit:
Added named credential and changed Account Factory

Commit:
Changed Factory class
